src/picongpu/include/particles/bremsstrahlung/Bremsstrahlung.tpp

```diff
@@ -0,0 +1,252 @@
+
+    float_X ionDensity = ionCount_norm.x() *
+        (static_cast<float_X>(particleToGrid::derivedAttributes::Counter().getUnit().x()) / CELL_VOLUME);
+
+    const float_X weighting = particle[weighting_];
+
+    const float_X c = SPEED_OF_LIGHT;
+    float3_X mom = particle[momentum_] / weighting;
+    const float_X momAbs = math::abs(mom);
+    float3_X mom_norm = mom / momAbs;
+
+    const float_X mass = frame::getMass<FrameType>();
+    const float_X momAbs_norm = momAbs / (mass * c);
+    const float_X Ekin = (math::sqrt(float_X(1.0) + momAbs_norm*momAbs_norm) - float_X(1.0)) * mass * c*c;
+    if(Ekin < electron::MIN_ENERGY)
+        return 0;
+
+    /* electron deflection */
+    const float_X zMin = float_X(1.0) / (float_X(M_PI)*float_X(M_PI));
+    const float_X zMax = float_X(1.0) / (electron::MIN_THETA*electron::MIN_THETA);
+    const float_X z = zMin + this->randomGen() * (zMax - zMin);
+    const float_X theta = math::rsqrt(z);
+    const float_X theta2 = theta*theta;
+    const float_X rutherfordCoeff = ELECTRON_CHARGE*ELECTRON_CHARGE / (float_X(4.0) * float_X(M_PI) * EPS0) * ion::TARGET_Z / Ekin;
+    const float_X denominator = (electron::MIN_THETA*electron::MIN_THETA + theta2);
+    const float_X scaledDeflectionDCS = float_X(1.0) / float_X(2.0) * (zMax - zMin) * rutherfordCoeff*rutherfordCoeff *
+        float_X(2.0) * float_X(M_PI) * theta2*theta2 / (denominator*denominator);
+    const float_X deflectionProb = ionDensity * c * DELTA_T * scaledDeflectionDCS;
+
+    if(this->randomGen() < deflectionProb)
+    {
+        mom = this->scatterByTheta(mom, theta);
+        mom_norm = mom / momAbs;
+    }
+
+    /* non-radiative Bremsstrahlung */
+    const float_X kappaCutoff = math::min(this->softPhotonsCutoff / Ekin, float_X(1.0));
+    const float_X stoppingPower = ionDensity * c * this->stoppingPowerFunctor(Ekin, kappaCutoff);
+    const float_X newEkin = math::max(Ekin - stoppingPower * DELTA_T, float_X(0.0));
+    const float_X newEkin_norm = newEkin / (mass * c*c);
+    const float_X newMomAbs = mass * c * math::sqrt(newEkin_norm*newEkin_norm + float_X(2.0) * newEkin_norm);
+    const float_X deltaMom = newMomAbs - momAbs;
+    particle[momentum_] = (mom + deltaMom * mom_norm) * weighting;
+
+    /* photon emission */
+    const float_X delta = this->randomGen();
+    const float_X kappa = math::pow(kappaCutoff, float_X(1.0) - delta);
+    const float_X scalingFactor = -math::log(kappaCutoff);
+    const float_X emissionProb = scalingFactor * ionDensity * c * DELTA_T * this->scaledSpectrumFunctor(Ekin, kappa);
+
+    // raise a warning if the emission probability is too high.
+    if(picLog::log_level & picLog::CRITICAL::lvl)
+    {
+        if(emissionProb > float_X(photon::SINGLE_EMISSION_PROB_LIMIT))
+        {
+            const float_X Ekin_SI = Ekin * UNIT_ENERGY;
+            printf("[Bremsstrahlung] warning: emission probability is too high: \
```

Comment:
@psychocoderHPC You are right - this is device code only. So this is fine!
Or does `fprintf(stderr, "...")` work?